src/plugins/intel_gpu/src/plugin/transformations/sink_reshape.cpp

```diff
@@ -0,0 +1,173 @@
+            size_t mismatch_count = 0;
+            if (in_ps.size() > out_ps.size()) {
+                for (size_t i = 0; i < out_ps.size(); ++i) {
+                    if (i + mismatch_count >= in_ps.size())
+                        return false;
+                    if (out_ps[i] != in_ps[i + mismatch_count]) {
+                        mismatch_count++;
+                    }
+                }
+                return mismatch_count == 1;
+            }
+            return true;
+        };
+        const auto reshape = ov::as_type_ptr<v1::Reshape>(output.get_node_shared_ptr());
+        return is_suitable_reshape(reshape) && is_suitable_parent(reshape->get_input_node_shared_ptr(0));
+    };
+
+    auto reshape_m = wrap_type<v1::Reshape>(reshape_predicate);
+    auto transpose_const_m = wrap_type<v0::Constant>();
+    auto transpose_m = wrap_type<v1::Transpose>({reshape_m, transpose_const_m});
+
+    ov::matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS](ov::pass::pattern::Matcher& m) {
+        const auto& pattern_map = m.get_pattern_value_map();
+        auto reshape = std::dynamic_pointer_cast<v1::Reshape>(pattern_map.at(reshape_m).get_node_shared_ptr());
+        if (!reshape || transformation_callback(reshape)) {
+            return false;
+        }
+
+        auto update_order = [](std::vector<uint16_t> original_order, const std::shared_ptr<v1::Reshape>& reshape_node) {
+            // Example. For this sequence, there is Reshape node which merges 2 consecutive dims into one
+            // order must be updated like permute is done before reshape
+            // [1,3,4,6] -> Reshape[1,3,24,1]-> permute(0,2,1) -> [1,24,3,1]
+            // updated order must be (0,2,3,1):
+            // dim with index=2 is split into 2 parts: 2 and 3
+            auto reshape_in_shape = reshape_node->get_input_partial_shape(0).to_shape();
+            auto reshape_out_shape = reshape_node->get_output_partial_shape(0).to_shape();
+            auto transformed_order = original_order;
+            ov::Shape new_shape(transformed_order.size());
+            const uint16_t merge_dim_idx = [&]() {
+                for (uint16_t i = 0; i < reshape_out_shape.size(); ++i) {
+                    if (reshape_in_shape[i] != reshape_out_shape[i])
+                        return i;
+                }
+                OPENVINO_THROW("same input/output for reshape node");
+            }();
+            auto insertIt = transformed_order.end();
+            for (auto it = transformed_order.begin(); it != transformed_order.end(); ++it) {
+                auto& elem = *it;
+                if (elem > merge_dim_idx) {
+                    elem++;
+                } else if (elem == merge_dim_idx) {
+                    insertIt = it + 1;
+                }
+            }
+            transformed_order.insert(insertIt, merge_dim_idx + 1);
+            // remove invalid orders
+            if (transformed_order.size() > reshape_in_shape.size()) {
```

Comment:
In reshape => transpose, at this phase (i.e., ngraph level, not cldnn graph), reshape output rank should be always same as transpose's input rank == permute order. 

Also I am wondering whther Reshape you are observing is actually 4d=>4d. 
I guess that it will be 4D => 3D and then expanded to 4D in the cldnn primitive creation step. Isn't it? 
That "expansion to 4D " is our legacy which is only applied to the static model, and it will be removed in some while (Vladimir was preparing that change) 
I think we don't need to care about different rank of permute order vs Reshape rank at this moment.

Reply:
em...yes,  the issue is found with 4D->3D case,  and the expansion is done in cldnn primitive. fixed.